### DISCUSSION TITLE
NetworkX Visualization Improvements

### DISCUSSION THREAD

**Discussion by mjschwenne** [networkx/networkx#7520] [Ideas]
NetworkX is an extreme popular library, and I would argue for a lot of good reasons. However, one area where we struggle is graph visualization. This is an extremely difficult problem due to how complex graphs are, even spawning dedicated tools like [Graphviz](https://graphviz.org/), [Gephi](https://gephi.org/) and [Cytoscape](https://cytoscape.org/). While it\'s not feasible to make NetworkX\'s visualization suite as expansive as these more specialized programs, I think there is ample room for improvement. In the post below, I\'ve outlined some of my thoughts on the current state of NetworkX\'s visualization tools, but I\'d appreciate input from other community members since I will be working this important topic this summer. I expect most of the work will be focused on the NetworkX API for graph visualization, rather then overhauling the visualization code itself, but as the need arises I will work on all parts of the visualization stack.

- After talking with some NetworkX users, it seems like people appreciate the fact that all of the node and edge attributes are stored internally on the graph object and does not require tracking extra indices or maps. However, the visualization API is one place were this is not true. Just take a look at the documentation for `draw_networkx` and you will see multiple parameters like `pos`, `node_size`,   `edge_color`, etc. which take maps or arrays that we use to describe how to plot an node or edge. Realistically, a node or edge should contain the attributes needed to visualize itself. It would be nice to store these attributes directly on the nodes and edges and have the drawing functions read them from the graph. There are a few considerations before committing to this approach. One benefit of this approach is that `set_node_attributes` already lets us set one value across all nodes and `set_edge_attributes` works in the same manner. However, it will let all attributes be heterogeneous. A lot of the `draw_networkx` parameters allow for arrays, but things like font size, font family, alpha, etc. do not. I don\'t think this will be a major  problem, but we will have to expand the code to account for this difference. If we take this approach, the layout functions should default to setting the node positions as an attribute rather than returning a dictionary of positions.

- What is the difference between `nx.draw` and `nx.draw_networkx`? Not much, `draw` is basically a wrapper around `draw_networkx` that sets up the matplotlib    axes. Besides having names that I think are too similar, I wonder if we  could combine some of `draw_networkx`, `draw_networkx_edges`, `draw_networkx_labels` and  `draw_networkx_edge_labels`. These functions are important for sequencing the additions to the canvas, but how often would somebody want to draw just the edge labels? I don\'t think we should remove the functionality itself, but perhaps a parameter to `nx.draw` such as `nodes=None` would suppress the drawing of the nodes?

-   Resolve some of the visualization issues. I think that these issues
    are resolvable:
    -   #4525 - possibly involving the creation of a graph artist for
        matplotlib.
    -   #4535 - The results of the labels should be consistent, although
        at this point I\'m not sure which drawing method needs to be
        changed.
    -   #5206
    -   #5609
    -   #6877
    -   #7054

    Issues that seem particularly difficult:
    -   #4601 - Not really sure if a solution is possible here, but the
        inexact solution proposed by the original poster (which I don\'t
        believe has been implemented) does yield better results.
    -   #4885

- Since the 3.0 release, NetworkX has supported backends. While these have been    used successfully for algorithms, we should be able to use them for graph    visualization as well. Using the \"graph describes how it looks\" philosophy, I see two ways the dispatching system could be useful for visualization.
    
    - Convert our existing code for pygraphviz / pydot into a backend for the standard `draw` function. There may be ways to use this to enhance the interactivity in matplotlib plots as well.

    - Extend the visualization capabilities to use other backends, such as using
        [Plotly](https://plotly.com/python/), [Altair](https://altair-viz.github.io/index.html) or [Bokeh](https://docs.bokeh.org/en/3.3.0/index.html).

Realistically, this is probably more than one summer of work, but even accomplishing the first one or two bullet points would represent a significant improvement to the API for NetworkX graph visualizations in terms of ease-of-use and making it more consistent with the rest of the library.

I really am interested in feedback and input on how to best organize this work as I spend some more time familiarizing myself with the internals of the drawing functionality we already have.

**Comment by dschult** [comment]
[GraVe](https://github.com/networkx/grave) might be out of date, and was never fully developed by any means -- but it represents the perspective of a matplotlib expert's approach to making a networkx interface for drawing graphs. It was created during a meeting at BIDS in about a day with lots of discussions about interfaces. There might be good ideas there.  Both in the code and in the examples.

I didn't understand your statements about setting node/edge attributes in networkx:
>One benefit of this approach is that set_node_attributes already lets us set one value across all nodes and set_edge_attributes works in the same manner. However, it will let all attributes be heterogeneous.

I believe `set_node_attributes` and `set_edge_attributes` both work with the same interface, allowing us to set a single data value for all nodes, or inputting a dict keyed by node/edge to the value of interest. I believe these functions are fairly quick -- at least for the small-ish networks we are aiming for (<~=10K nodes???) so having the API allow specifying either a data-name string for data on the Graph or a dict of values or a function taking in the node/edge as input would be a nice flexible system. We can focus first on using data on the Graph -- but as you design it, try to centralize data handling helper functions so it's easy to expand for other data storage formats.

The various functions `draw`, `draw_networkx`, `draw_networkx_edges` were originally helper functions for `draw` that we broke out so people who want to do specialized stuff could construct what they wanted without rewriting the whole big function. This should probably be re-factored since it was constructed long ago and much has changed. But I encourage a design philosophy that aims for:
- make simple/quick drawing easy and still allow detailed complicated drawing to be possible.
- give enough lego-like tools to the user that they can put those pieces together to build whatever they want. The difficulty here is identifying what are the key lego-shapes for this problem.

The second bullet seems to discourage making a single monolithic function where we turn off functionality when we don't need it. But I don't think it rules it out. We want it to be easy to draw some nodes red and some blue with some edges red and others blue. Let's give people the tools to easily build that arrangement. That could be one monolithic function with side helper functions, or it could separate the Artist tool functionality by type of object with helper functions to gather those together in flexible ways.  It might be worth looking at other data-science drawing tools (seaborn?, matplotlib itself?) to see how their API handles this and whether that can be improved or imitated.

**Comment by mjschwenne** [comment]
I've been meaning to create a follow up post here as I've been writing responses and never posting them (Oops).

While Dan isn\'t convinced about one monolithic function for visualization, I think that it accomplishes a goal that he didn\'t state but other maintainers have expressed: that visualization isn\'t a primary aim for networkx so I'm weary about pushing the visualization capabilities too far beyond what we already have.

Using the current API, if I wanted to plot only the edges of a graph I could pass `nodelist=[]` and `edgelist=list(G.edges())` to get the same result. The only difference between these calls is that `draw_networkx_edges` leaves a border around the drawing.

``` python
pos = nx.spring_layout(G)
nx.draw(G, nodelist=[], pos=pos)
```
![image](https://github.com/user-attachments/assets/46e64b18-d128-4bf4-9f1c-af51a18fcf29)

``` python
nx.draw_networkx_edges(G, pos)
```
![image](https://github.com/user-attachments/assets/d49e9986-8668-4d8f-a910-af77f19e2b40)

Ok, perhaps we don\'t want solely one monolithic function, but I\'m unconvinced that the current other matplotlib helper functions are
really useful given the flexibility that `nx.draw` already has. What *would* be useful is some helper functions which can help get this data
on the graph object and then off of it again.

## Visualization Parameters

A lot of these attributes are opt-in. For example, if you don\'t provide a label attribute on a node, it will be displayed without a label. But
what should happen if you give a label but no font size, font color, etc. In that case we could either error or infer a default. Throwing an
error would make the interface too brittle and put too much work on the user to all attributes they they didn\'t have to do in the old
interface. Flexibility is already one of our strengths and we should focus on keeping that.

What I, in the abstract, want is a function where each of these attributes is a parameter. The values of that parameter can be:

-   Some default value applied to all nodes or edges
-   The name of an edge or node attribute to read from the graph
-   A function which is called on each node/edge and returns the  attribute for that node/edge.

However, there are several issues with this.

-   How do we handle incomplete node/edge attributes? My thought is that we allow all attributes for which a default value exists to be
    incomplete. For any value which is missing from the node/edge, we plug in the default.

-   How do we tell the difference between attribute name strings and default value strings? We\'ll have to check the graph. I think that
    it follows from the pervious bullet that if the value exists as an attribute name on any of the nodes or edges of the graph (restricted
    to respect the `nodelist` and `edgelist`) we interpret the value as an attribute name. While it\'s possible that this could cause conflicts.  
    Consider a graph with visualization attributes and attributes from a graph analysis being performed. If I want the default font size to be 12 and not 10, but for some reason 12 is the name of an attribute on the graph. What happens?

-   If the input is a function, we should call that function with the node/edge and all of it\'s attributes. If there are multiple function parameters which depend on each other (i.e. to compute font size I need to know the node color but that\'s also computed via a function) how do we order the function calls? I think this problem is beyond the scope of this work and the user should be responsible for getting the final data onto the graph.

    After all, this does the same thing.

    ``` python
    nx.set_node_attributes(G, {n: f(n) for n in G.nodes()}, "font_color")
    ```

-   What about the label sub-attributes. I\'d like to not have to pass  them all individually in a call to draw. How about `label` takes a
    string or dict and if it\'s a dict we give all of the sub-attributes that way?

## Dispatching

We want this work to allow for dispatching to reach the visualization tools. There are a few things that we can do to facilitate on the
NetworkX API side of things.

-   Return values from `nx.draw`: Some visualization tools like bokeh, graph-tool, plotly and holoviews allow for the creation of
    interactive plots. If the user is able to move node positions around or otherwise change the graph, we want to be able to handle those
    changes. I think the simple solution here is to return the post-visualization graph. If changes are made, that will be different from the pre-visualization graph, but for static visualizations it will always be the same and the user can disregard the return value.

But a problem we have with dispatching for visualization is the matplotlib specific visualization parameters. Those are:

-   `nx.draw_networkx_nodes`
    -   `ax`: The matplotlib axes to plot on.
    -   `cmap`: A matplotlib color map to pull for node colors.
    -   `vmin` and `vmax`: supports clipping the low and high parts of
        the color map.
    -   `hide_ticks`: Weather to hide ticks and tick labels on the
        border of the plot.
-   `nx.draw_networkx_edges`
    -   `ax`
    -   `edge_cmap`
    -   `edge_vmin` and `edge_vmax`
    -   `connectionstyle`: Controls how curved the arc is.
    -   `hide_ticks`
-   `nx.draw_networkx_labels`
    -   `ax`
    -   `bbox`: A bounding box for the label.
    -   `hide_ticks`
-   `nx.draw_edge_labels`
    -   `ax`
    -   `bbox`
    -   `connectionstyle`
    -   `hide_ticks`

Some of these can be handled via existing backend kwarg logic, specifically the `ax` and `hide_ticks` parameters, but the colors, bounding box and connection style parameters will need non-matplotlib specific options.

The colormap could be handled via a helper function which takes the map and computes what the output color is and applies it to a new node/edge attribute so that we have hex codes for colors. This would also allow matplotlib color maps to be using (indirectly) with other visualization backends when those are ready.

I\'m honestly not sure what to do about the bounding box or connection style parameters.
